stop an app deployed to mvm

etanshaul opened the pull request:
Fixes #453

In progress. Functionality works, just needs some cleanup.

etanshaul commented:
@patflynn @chanseokoh @meltsufin, you guys can start taking a look at this

patflynn commented on google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties:
```diff
@@ -45,6 +45,8 @@ appengine.appyaml.location.label=app.yaml path\:
 appengine.dockerfile.location.browse.button=Browse for Your Dockerfile Location
 appengine.deployment.error.with.code=Deployment failed with exit code: {0}\nPlease make sure that you are using the latest version of the Google Cloud SDK.\nRun ''gcloud components update'' to update the SDK. (See: https://cloud.google.com/sdk/gcloud/reference/components/update.)
 appengine.deployment.error.during.execution=Deployment failed due to an unexpected error during the deployment execution phase.
+appengine.stopapp.error.during.execution=Failed to stop application due to an unexpected error during the execution phase.
```
appengine.stop.module.version.execution.error?

I'm just wondering if we might end up supporting module or app level stop.

etanshaul replied:
How about: appengine.stop.modules.version.execution.error

All I changed from your suggestion is making module plural.  It reads a little awkwardly but it makes it consistent with the underlying gcloud command. The stop could potentially act on multiple modules for a multimodule deployment.

patflynn replied:
+1

On Fri, Apr 1, 2016 at 2:04 PM, Etan Shaul <notifications@github.com> wrote:

> In
> google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/571#discussion_r58243714>
> :
>
> > @@ -45,6 +45,8 @@ appengine.appyaml.location.label=app.yaml path\:
> >  appengine.dockerfile.location.browse.button=Browse for Your Dockerfile Location
> >  appengine.deployment.error.with.code=Deployment failed with exit code: {0}\nPlease make sure that you are using the latest version of the Google Cloud SDK.\nRun ''gcloud components update'' to update the SDK. (See: https://cloud.google.com/sdk/gcloud/reference/components/update.)
> >  appengine.deployment.error.during.execution=Deployment failed due to an unexpected error during the deployment execution phase.
> > +appengine.stopapp.error.during.execution=Failed to stop application due to an unexpected error during the execution phase.
>
> How about: appengine.stop.modules.version.execution.error
>
> All I changed from your suggestion is making module plural. It reads a
> little awkwardly but it makes it consistent with the underlying gcloud
> command. The stop could potentially act on multiple modules for a
> multimodule deployment.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/571/files/f30847bf0cdb604ecbec0439a415f0fd6c2328d1#r58243714>
>

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineAction.java:
```diff
@@ -0,0 +1,144 @@
+      @NotNull GeneralCommandLine commandLine,
+      @NotNull File appDefaultCredentialsPath,
+      @NotNull ProcessListener listener) throws ExecutionException {
+    consoleLogLn("Executing: " + commandLine.getCommandLineString());
```
should we be logging the command string before we add the base parameteres?

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineAction.java:
```diff
@@ -0,0 +1,144 @@
+    // Common command line settings
+    commandLine.addParameter("--project=" + appEngineHelper.getProjectId());
+    commandLine
+        .addParameter("--credential-file-override=" + appDefaultCredentialsPath.getAbsolutePath());
```
why involve the subclass in generating the credentials at all?

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/util/AppEngineUtil.java:
```diff
@@ -0,0 +1,42 @@
+   * @return generated version
+   */
+  public static String generateVersion() {
+    String versionPattern = "yyyyMMdd't'kms";
```
I wonder if cloud SDK always uses the local timezone?  Feels like that would be bad for developers deploying to the same project across timezones.

etanshaul replied:
Great catch. I'll ask that question on the cloud sdk thread.

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/util/AppEngineUtil.java:
```diff
@@ -0,0 +1,42 @@
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+public class AppEngineUtil {
```
class javadoc is required.

etanshaul commented:
Updated with comments so far.

Moved all credential checking to the superclass.

meltsufin commented:
From the UX point of view, I find it a little odd how when you click the undeploy button, the module just instantly disappears and there is no confirmation that the undeployment was successful. Contrast that with the deploy action which takes a while, and you see the process happening in the console. I think it would be nice to bring up a popup confirming the success of the action.

etanshaul commented:
I agree. Or instead of a popup if we could get the console output to remain visible showing some success message indicating that the app was undeployed.

meltsufin commented:
Yes, either that, or maybe we can even hook into the IJ Event Log somehow? Right now it only seems to report log.error messages in the Event Log.

etanshaul commented:
The event log doesn't feel like the right place for a success message. Ideally it would be in the console view but this doesn't look possible because of how IJ deletes the deployment after a successful undeploy (and the console goes away with it). I'm exploring using a dialog here.

etanshaul commented:
@meltsufin I added a success dialog to the stop process. Let me know what you think.

meltsufin commented on google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties:
```diff
@@ -45,6 +45,10 @@ appengine.appyaml.location.label=app.yaml path\:
 appengine.deployment.error.during.execution=Deployment failed due to an unexpected error during the deployment execution phase.
+appengine.stop.modules.version.execution.error=Failed to stop application due to an unexpected error during the execution phase.
+appengine.stop.modules.version.execution.error.with.code=Stop application failed with exit code: {0}
+appengine.stop.modules.version.success.dialog.message=Application successfully undeployed.
```
I guess the correct terminology would be "stopped", not "undeployed". Also, maybe specify which application in the message?
"App Engine application {0} has successfully stopped." or something like that.

meltsufin commented:
Maybe it then also makes sense to have a similar dialog for completion of deployment as well?

etanshaul commented:
Yea. I think so.

etanshaul commented:
- Added successful deployment dialog
- Added an undeployment confirmation dialog. Before, the application stop was immediately executed when clicking the button. Now the user can cancel out of it.

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineAction.java:
```diff
@@ -0,0 +1,155 @@
+    this.loggingHandler = loggingHandler;
+    this.appEngineHelper = appEngineHelper;
+
+    credentialsPath = createApplicationDefaultCredentials();
```
we probably want to minimize the window between when we create the credential file and we execute.  Could we put this this in executeProvess?

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineAction.java:
```diff
@@ -0,0 +1,155 @@
+
+    GoogleLoginState googleLoginState;
+    if (projectUser != null) {
+      googleLoginState = projectUser
```
this can fit on one line.

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineAction.java:
```diff
@@ -0,0 +1,155 @@
+  private static final String GCLOUD_USER_TYPE = "authorized_user";
+
+  @VisibleForTesting
+  protected File createApplicationDefaultCredentials() {
```
@Nullable

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineDeployAction.java:
```diff
@@ -0,0 +1,248 @@
+                if (doStop == Messages.YES) {
+                  stop(callback);
+                } else {
+                  callback.errorOccurred("Undeploy Canceled");
```
so this leaves it in a state where you can still click 'undeploy'?

etanshaul replied:
yes. specifically invoking errorOccured is necessary for that.

patflynn replied:
ok. 'Undeploy Cancelled' should be localized. 

Also did you look at showOkCancelDialog?

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineDeployAction.java:
```diff
@@ -0,0 +1,248 @@
+          private void stop(@NotNull UndeploymentTaskCallback callback) {
+            Set<String> modulesToSop;
+            try {
+              modulesToSop = parseDeployOutputToModuleList(deployOutput.toString());
```
Can you file a bug to update this code to use the module name configured (once we have module name configuration)?

etanshaul replied:
Sure. I think updating this code will be part of whatever PR includes module name configuration.

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineDeployAction.java:
```diff
@@ -0,0 +1,248 @@
+    @Override
+    public void processTerminated(final ProcessEvent event) {
+      if (event.getExitCode() == 0) {
+        callback.succeeded(new DeploymentRuntime() {
```
can we define this DeploymentRuntime somewhere else?  It distracts from the flow.

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineDeployAction.java:
```diff
@@ -0,0 +1,248 @@
+          @Override
+          public void run() {
+            Messages.showMessageDialog(
+                GctBundle.message("appengine.deployment.success.dialog.message"),
```
does the user have to dismiss this?  I think it's kind of not IntelliJ style to have a dialog for a successful run configuration

etanshaul replied:
Yes, the user has to dismiss it. The reason I created this one is to keep it consistent with the stop flow which now has a "successful stop" dialog. This dialog is not completely necessary since the user can see a message of succesfull deployment in the console..

patflynn replied:
Yeah.  I can see why we added it to the stop flow. But I'd prefer we left
it out of a successful deployment.

On Fri, Apr 1, 2016 at 5:13 PM, Etan Shaul <notifications@github.com> wrote:

> In
> google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineDeployAction.java
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/571#discussion_r58268442>
> :
>
> > +            ProgressManager.getInstance()
> > +                .run(new Task.Backgroundable(project, "Stop App Engine", true,
> > +                    null) {
> > +                  @Override
> > +                  public void run(@NotNull ProgressIndicator indicator) {
> > +                    ApplicationManager.getApplication().invokeLater(appEngineStopAction);
> > +                  }
> > +                });
> > +          }
> > +        });
> > +
> > +        SwingUtilities.invokeLater(new Runnable() {
> > +          @Override
> > +          public void run() {
> > +            Messages.showMessageDialog(
> > +                GctBundle.message("appengine.deployment.success.dialog.message"),
>
> Yes, the user has to dismiss it. The reason I created this one is to keep
> it consistent with the stop flow which now has a "successful stop" dialog.
> This dialog is not completely necessary since the user can see a message of
> succesfull deployment in the console..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-intellij/pull/571/files/48df9093626829470ef71e2e3c3b392e63ad2108#r58268442>
>

etanshaul replied:
Sounds good to me.

meltsufin replied:
Given that we now also have a confirmation before undeploy action, maybe we don't need another one after, if the action is successful. That way neither deploy nor undeploy would have a success confirmation dialog.

etanshaul replied:
That idea also has the nice benefit of getting rid of "useless" informational dialogs.

patflynn commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineStopAction.java:
```diff
@@ -0,0 +1,109 @@
+            GctBundle.message("appengine.stop.modules.version.execution.error.with.code", event.getExitCode()));
+      }
+
+      deleteCredentials();
```
I'm wondering if we want to include deleteCredentials in a finally clause to ensure it always happens.

etanshaul replied:
I think that makes sense

etanshaul commented:
Comments addressed.

I left the undeployment success info dialog since I'm on the fence about it. Without it, the user has no indication that the stop was successful.

meltsufin commented:
LGTM

patflynn commented:
LGTM>
